File: npc_sim/decisions/action_library.py
# Copyright 2025-2026 Sadık Abdusselam Albayrak
# Licensed under the Apache License, Version 2.0
"""Registry of all available actions."""

from __future__ import annotations
from npc_sim.decisions.action import IAction


class ActionLibrary:
    """Registry of all available actions in the simulation."""

    def __init__(self):
        self._actions: list[IAction] = []

    def register(self, action: IAction) -> None:
        if action is not None:
            self._actions.append(action)

    def get_all(self) -> list[IAction]:
        return self._actions

    def __repr__(self) -> str:
        return f"[ActionLibrary] {len(self._actions)} action(s)"

File: Runtime/Decisions/ActionLibrary.cs
using System;
using System.Collections.Generic;

namespace ForgeProject.Sim
{
    /// <summary>
    /// Registry of all <see cref="IAction"/> implementations available to the simulation.
    /// Populated at startup; read-only during simulation ticks for thread safety.
    /// </summary>
    public sealed class ActionLibrary
    {
        private readonly List<IAction> _actions = new();

        // ─── Registration ─────────────────────────────────────────────────────────

        /// <summary>Registers an action. Duplicate ActionIds are ignored.</summary>
        public void Register(IAction action)
        {
            if (action == null) return;
            foreach (var a in _actions)
                if (a.ActionId == action.ActionId) return;
            _actions.Add(action);
        }

        // ─── Queries ──────────────────────────────────────────────────────────────

        /// <summary>All registered actions.</summary>
        public IReadOnlyList<IAction> GetAll() => _actions;

        /// <summary>Actions matching the given <see cref="IAction.ActionType"/>.</summary>
        public List<IAction> GetByType(string actionType)
        {
            var result = new List<IAction>();
            foreach (var a in _actions)
                if (a.ActionType.Equals(actionType, StringComparison.OrdinalIgnoreCase))
                    result.Add(a);
            return result;
        }

        public override string ToString() => $"[ActionLibrary] {_actions.Count} action(s)";
    }
}
